server/schedule/operator/builder.go

```diff
@@ -0,0 +1,596 @@
+	return p.addOrPromote == nil && p.remove == nil
+}
+
+func (b *Builder) peerPlan() stepPlan {
+	// Replace has the highest priority because it does not change region's
+	// voter/learner count.
+	if p := b.planReplace(); !p.empty() {
+		return p
+	}
+	if p := b.planRemovePeer(); !p.empty() {
+		return p
+	}
+	if p := b.planAddPeer(); !p.empty() {
+		return p
+	}
+	if p := b.planPromotePeer(); !p.empty() {
+		return p
+	}
+	return stepPlan{}
+}
+
+func (b *Builder) planReplace() stepPlan {
+	var best stepPlan
+	for _, i := range b.toRemove.IDs() {
+		r := b.toRemove.Get(i)
+		for _, j := range b.toAdd.IDs() {
+			if i == j {
+				// Special case: remove voter and add learner on same store.
+				// Cannot use replace. Need first removePeer then addPeer.
+				continue
+			}
+			a := b.toAdd.Get(j)
+			if r.GetIsLearner() == a.GetIsLearner() {
+				best = b.planReplaceDetail(best, stepPlan{addOrPromote: a, remove: r})
+			}
+		}
+		if !r.GetIsLearner() {
+			for _, j := range b.toPromote.IDs() {
+				a := b.toPromote.Get(j)
+				best = b.planReplaceDetail(best, stepPlan{addOrPromote: a, remove: r})
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planReplaceDetail(best, next stepPlan) stepPlan {
+	// Replace peer with the nearest one.
+	if mb, mc := b.labelMatch(best), b.labelMatch(next); mb > mc {
+		return best
+	} else if mb < mc {
+		best = stepPlan{}
+	}
+
+	// Brute force all possible leader combinations to find the best plan.
+	for _, leader1 := range b.currentPeers.IDs() {
+		if !b.allowLeader(b.currentPeers.Get(leader1)) {
+			continue
+		}
+		next.leader1 = leader1
+		for _, leader2 := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader2)) {
+				next.leader2 = leader2
+				best = b.comparePlan(best, next)
+			}
+		}
+		if b.allowLeader(next.addOrPromote) {
+			next.leader2 = next.addOrPromote.GetStoreId()
+			best = b.comparePlan(best, next)
+		}
+	}
+	return best
+}
+
+func (b *Builder) planRemovePeer() stepPlan {
+	var best stepPlan
+	for _, i := range b.toRemove.IDs() {
+		r := b.toRemove.Get(i)
+		for _, leader := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader)) {
+				best = b.comparePlan(best, stepPlan{remove: r, leader2: leader})
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planAddPeer() stepPlan {
+	var best stepPlan
+	for _, i := range b.toAdd.IDs() {
+		a := b.toAdd.Get(i)
+		for _, leader := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader)) {
+				best = b.comparePlan(best, stepPlan{addOrPromote: a, leader1: leader})
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planPromotePeer() stepPlan {
+	for _, i := range b.toPromote.IDs() {
+		p := b.toPromote.Get(i)
+		return stepPlan{addOrPromote: p, leader1: b.currentLeader}
+	}
+	return stepPlan{}
+}
+
+// Pick the better plan from 2 candidates.
+func (b *Builder) comparePlan(best, next stepPlan) stepPlan {
+	if best.empty() {
+		return next
+	}
+	fs := []func(stepPlan) int{
```

Comment:
Could we add a `preferAddTargetLeader` to make the target leader be added as early as possible? The reason is the same as `preferTargetLeader`.

Reply:
It should be `preferAddOrPromoteTargetLeader`.

Reply:
Good idea.

Reply:
fixed.

Reply:
It's better to add some comments to explain why the order should be like this.

Reply:
good idea @rleungx

Reply:
fixed